cirq-core/cirq/circuits/circuit_operation.py

```diff
@@ -110,84 +110,123 @@ class CircuitOperation(ops.Operation):
+        if not isinstance(self._circuit, circuits.FrozenCircuit):
+            raise TypeError(f'Expected circuit of type FrozenCircuit, got: {type(self._circuit)!r}')
 
         # Ensure that the circuit is invertible if the repetitions are negative.
-        if isinstance(self.repetitions, float):
-            if math.isclose(self.repetitions, round(self.repetitions)):
-                object.__setattr__(self, 'repetitions', round(self.repetitions))
-        if isinstance(self.repetitions, INT_CLASSES):
-            if self.repetitions < 0:
+        self._repetitions = repetitions
+        self._repetition_ids = None if repetition_ids is None else list(repetition_ids)
+        self._use_repetition_ids = use_repetition_ids
+        if isinstance(self._repetitions, float):
+            if math.isclose(self._repetitions, round(self._repetitions)):
+                self._repetitions = round(self._repetitions)
+        if isinstance(self._repetitions, INT_CLASSES):
+            if self._repetitions < 0:
                 try:
-                    protocols.inverse(self.circuit.unfreeze())
+                    protocols.inverse(self._circuit.unfreeze())
                 except TypeError:
                     raise ValueError('repetitions are negative but the circuit is not invertible')
 
             # Initialize repetition_ids to default, if unspecified. Else, validate their length.
-            loop_size = abs(self.repetitions)
-            if not self.repetition_ids:
-                object.__setattr__(self, 'repetition_ids', self._default_repetition_ids())
-            elif len(self.repetition_ids) != loop_size:
+            loop_size = abs(self._repetitions)
+            if not self._repetition_ids:
+                self._repetition_ids = self._default_repetition_ids()
+            elif len(self._repetition_ids) != loop_size:
                 raise ValueError(
                     f'Expected repetition_ids to be a list of length {loop_size}, '
-                    f'got: {self.repetition_ids}'
+                    f'got: {self._repetition_ids}'
                 )
-        elif isinstance(self.repetitions, sympy.Expr):
-            if self.repetition_ids is not None:
+        elif isinstance(self._repetitions, sympy.Expr):
+            if self._repetition_ids is not None:
                 raise ValueError('Cannot use repetition ids with parameterized repetitions')
         else:
             raise TypeError(
                 f'Only integer or sympy repetitions are allowed.\n'
-                f'User provided: {self.repetitions}'
+                f'User provided: {self._repetitions}'
             )
 
+        # Disallow qid mapping dimension conflicts.
+        self._qubit_map = qubit_map or {}
+        for q, q_new in self._qubit_map.items():
+            if q_new.dimension != q.dimension:
+                raise ValueError(f'Qid dimension conflict.\nFrom qid: {q}\nTo qid: {q_new}')
+
+        self._measurement_key_map = measurement_key_map or {}
```

Comment:
Defensive copy for the dicts? Or comment in the docstring to state no defensive copy will be made?

Reply:
This is a good call, as neither should be large enough for copy to be an issue and the dict must not change. Technically a breaking change, but that's par for the course on this PR.